### PR TITLE
GPII-3146: Persistent volumes for CouchDB

### PR DESCRIPTION
**Pull request by natarajaya**
This is default chart settings: allow Kubernetes to create volumes for us, use "standard" storage class.

### CODE REVIEW

**Comment by mrtyler** [comment]
LGTM

**Comment by amatas** [comment]
LGTM :+1:

File: gcp/modules/couchdb/main.tf [hunk 1]
```diff
@@ -2,10 +2,64 @@ terraform {
...
+provider "google" {
+  project     = "${var.project_id}"
+  credentials = "${var.serviceaccount_key}"
+}
+
+resource "google_compute_disk" "couchdb" {
+  count = "${var.env == "dev" ? var.couchdb_dev_replicas : var.couchdb_prd_replicas}"
+  name  = "couchdb-${count.index}"
+  type  = "pd-ssd"
+  size  = 10
```
Comment (mrtyler) [inline gcp/modules/couchdb/main.tf]: This is set to 5 on the AWS side. Why did you change it here?

File: gcp/modules/couchdb/main.tf [hunk 2]
```diff
@@ -2,10 +2,64 @@ terraform {
...
+variable "serviceaccount_key" {}
+
+provider "google" {
+  project     = "${var.project_id}"
+  credentials = "${var.serviceaccount_key}"
+}
+
+resource "google_compute_disk" "couchdb" {
+  count = "${var.env == "dev" ? var.couchdb_dev_replicas : var.couchdb_prd_replicas}"
+  name  = "couchdb-${count.index}"
```
Comment (mrtyler) [inline gcp/modules/couchdb/main.tf]: Is the PV `couchdb-0` guaranteed to be associated with the couchdb Pod `couchdb-0`? If not, I fear this is going to be pretty confusing.

The AWS side names PVs based on their AZ, which I think is helpful. Not sure if it's possible here.

File: gcp/modules/couchdb/main.tf [hunk 3]
```diff
@@ -2,10 +2,64 @@ terraform {
...
+  type  = "pd-ssd"
+  size  = 10
+  zone  = "us-central1-a"
```
Comment (mrtyler) [inline gcp/modules/couchdb/main.tf]: I guess this is related to GPII-3194?

What do you think this code might look like when we have more than one AZ to worry about? Might it help with my question above about how PVs are named?

File: gcp/modules/couchdb/main.tf [hunk 4]
```diff
@@ -2,10 +2,64 @@ terraform {
...
+  provisioner "local-exec" {
+    command = <<EOF
+      echo "${join("", data.template_file.couchdb_pvs.*.rendered)}" | kubectl create -f -
+    EOF
+  }
+
+  provisioner "local-exec" {
+    when    = "destroy"
+    command = <<EOF
+      sleep 30 # Since there is no way to establish a proper dependency for module, we need to give couchdb helm release time to uninstall
```
Comment (mrtyler) [inline gcp/modules/couchdb/main.tf]: This seems kind of brittle. Can we solve this by 1) moving the PV management code to its own module and 2) using terragrunt dependencies so that the PVs and the couchdb chart are installed/uninstalled in the correct order?

File: gcp/modules/couchdb/resources/pv.yaml [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+  annotations:
+    backup.kubernetes.io/deltas: ${env == "prd" ? "PT5M PT60M PT24H P7D P52W" : (env == "stg" ? "PT15M PT60M PT4H PT24H P7D" : "PT5M PT15M PT45M")}
+spec:
+  capacity:
+    storage: 10Gi
+  accessModes:
+    - ReadWriteOnce
+  gcePersistentDisk:
+    pdName: couchdb-${index}
+  storageClassName: couchdb
```
Comment (mrtyler) [inline gcp/modules/couchdb/resources/pv.yaml]: Where are the properties for `storageClassName: couchdb` configured?

File: gcp/modules/templater/values/couchdb.yaml [hunk 6]
```diff
@@ -1,6 +1,12 @@
-# clusterSize is the initial size of the CouchDB cluster.
-clusterSize: 2
+clusterSize: ${env == "dev" ? 2 : 3}
 
 adminUsername: ${couchdb_admin_username}
 adminPassword: ${couchdb_admin_password}
 cookieAuthSecret: ${couchdb_secret}
+
+persistentVolume:
```
Comment (mrtyler) [inline gcp/modules/templater/values/couchdb.yaml]: I don't understand what this stanza does. Can you tell me more?

File: gcp/modules/couchdb/resources/pv.yaml [hunk 7]
```diff
@@ -0,0 +1,16 @@
+---
+apiVersion: v1
+kind: PersistentVolume
+metadata:
+  namespace: gpii
+  name: couchdb-${index}
+  annotations:
+    backup.kubernetes.io/deltas: ${env == "prd" ? "PT5M PT60M PT24H P7D P52W" : (env == "stg" ? "PT15M PT60M PT4H PT24H P7D" : "PT5M PT15M PT45M")}
```
Comment (mrtyler) [inline gcp/modules/couchdb/resources/pv.yaml]: I think these values also belong under `live/`.

File: gcp/modules/couchdb/main.tf (hunk 1 above)
Comment (natarajaya) [reply gcp/modules/couchdb/main.tf]: You can not create PDs smaller than 10GB for most storage classes in GCP.

File: gcp/modules/couchdb/main.tf [hunk 8]
```diff
@@ -2,10 +2,64 @@ terraform {
   backend "gcs" {}
 }
 
+variable "couchdb_prd_replicas" {
```
[the remark this answers is not in the document]
Comment (natarajaya) [reply gcp/modules/couchdb/main.tf]: Makes sense!

File: gcp/modules/couchdb/main.tf (hunk 2 above)
Comment (natarajaya) [reply gcp/modules/couchdb/main.tf]: I agree this can be confusing, let me investigate possible alternatives.

File: gcp/modules/templater/values/couchdb.yaml (hunk 6 above)
Comment (natarajaya) [reply gcp/modules/templater/values/couchdb.yaml]: Check https://github.com/gpii-ops/gpii-infra/blob/master/common/charts/couchdb/values.yaml#L28

**Comment by natarajaya** [comment]
@mrtyler I applied fixes following our discussion about K8s-managed volumes vs Terraform-managed.

File: gcp/modules/templater/values/couchdb.yaml (hunk 6 above)
Comment (mrtyler) [reply gcp/modules/templater/values/couchdb.yaml]: Ok

File: gcp/modules/couchdb/main.tf [hunk 9]
```diff
@@ -64,3 +68,39 @@ resource "null_resource" "couchdb_finish_cluster" {
...
+  depends_on = ["module.couchdb"]
+
+  triggers = {
+    couchdb_replicas = "${var.couchdb_replicas}"
+    backup_deltas    = "${var.backup_deltas}"
+  }
+
+  provisioner "local-exec" {
+    command = <<EOF
+      # We have to rely on jq for filtering, since it is not possible to use regexp in jsonpath:
```
[the remark this answers is not in the document]
Comment (natarajaya) [reply gcp/modules/couchdb/main.tf]: Unfortunately, yes. There is no way to filter output using only `jsonpath`.

[the remark this answers is not in the document]
Comment (mrtyler) [reply gcp/modules/couchdb/main.tf]: Ok. I don't understand why the comment is talking about regexp though? Maybe you can clarify it?

File: gcp/modules/couchdb/main.tf [hunk 10]
```diff
@@ -64,3 +68,39 @@ resource "null_resource" "couchdb_finish_cluster" {
...
+}
+
+resource "null_resource" "couchdb_destroy_pvcs" {
+  depends_on = ["module.couchdb"]
+
+  provisioner "local-exec" {
+    when = "destroy"
+    command = <<EOF
+      for PVC in $(kubectl get pvc --namespace ${var.release_namespace} -o json | jq --raw-output '.items[] | select(.metadata.name | startswith("database-storage-couchdb")) | .metadata.name'); do
+        kubectl --namespace ${var.release_namespace} delete pvc $PVC
```
[the remark this answers is not in the document]
Comment (mrtyler) [reply gcp/modules/couchdb/main.tf]: Good call on --ignore-not-found.

Orphaned resources are a general problem (on the AWS side, Snapshots from dev environments are the biggest offender) that we should solve generally, so this response makes sense to me.

File: gcp/live/dev/k8s/gpii/couchdb/terraform.tfvars [hunk 11]
```diff
@@ -18,3 +18,7 @@ terragrunt = {
 }
 
 # ↓ Module configuration (empty means all default)
+
+couchdb_replicas = 2
+backup_deltas = "PT5M PT15M PT45M"
```
Comment (mrtyler) [inline gcp/live/dev/k8s/gpii/couchdb/terraform.tfvars]: Where are the stg (and prd) values?

File: gcp/modules/couchdb/main.tf (hunk 9 above)
[the remark this answers is not in the document]
Comment (natarajaya) [reply gcp/modules/couchdb/main.tf]: Changed for `We have to rely on jq for filtering, since it is not possible to use regexp or string functions in kubectl jsonpath`

File: gcp/live/dev/k8s/gpii/couchdb/terraform.tfvars (hunk 11 above)
Comment (natarajaya) [reply gcp/live/dev/k8s/gpii/couchdb/terraform.tfvars]: We don't have prd in `live` yet, and what we have for stg is seriously outdated. Do you think I should work on this here?

Comment (mrtyler) [reply gcp/live/dev/k8s/gpii/couchdb/terraform.tfvars]: Yes.

File: gcp/modules/couchdb/main.tf [hunk 12]
```diff
@@ -64,3 +68,39 @@ resource "null_resource" "couchdb_finish_cluster" {
...
+  triggers = {
+    couchdb_replicas = "${var.couchdb_replicas}"
+    backup_deltas    = "${var.backup_deltas}"
+  }
+
+  provisioner "local-exec" {
+    command = <<EOF
+      # We have to rely on jq for filtering, since it is not possible to use regexp or string functions in kubectl jsonpath:
+      # https://github.com/kubernetes/kubernetes/issues/61406
+      for PV in $(kubectl get pv -o json | jq --raw-output '.items[] | select(.spec.claimRef.name | startswith("database-storage-couchdb")) | .metadata.name'); do
```
[the remark this answers is not in the document]
Comment (mrtyler) [reply gcp/modules/couchdb/main.tf]: Definitely.

File: gcp/modules/couchdb/main.tf (hunk 9 above)
[the remark this answers is not in the document]
Comment (mrtyler) [reply gcp/modules/couchdb/main.tf]: I still don't understand why this comment talks about regexp. Can you explain that part (or remove the reference, since I don't see any regular expressions anywhere in this PR)?

File: gcp/modules/couchdb/main.tf (hunk 12 above)
[the remark this answers is not in the document]
Comment (natarajaya) [reply gcp/modules/couchdb/main.tf]: PV is a storage resource in the cluster, and resources are not bound to namespaces:
```
/project # kubectl get pv
NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS    CLAIM                                     STORAGECLASS   REASON    AGE
pvc-21aeec86-896a-11e8-a157-42010a800006   10Gi       RWO            Delete           Bound     gpii/database-storage-couchdb-couchdb-0   standard                 19h
pvc-21b3c7f5-896a-11e8-a157-42010a800006   10Gi       RWO            Delete           Bound     gpii/database-storage-couchdb-couchdb-1   standard                 19h

/project # kubectl --namespace gpii get pv
NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS    CLAIM                                     STORAGECLASS   REASON    AGE
pvc-21aeec86-896a-11e8-a157-42010a800006   10Gi       RWO            Delete           Bound     gpii/database-storage-couchdb-couchdb-0   standard                 19h
pvc-21b3c7f5-896a-11e8-a157-42010a800006   10Gi       RWO            Delete           Bound     gpii/database-storage-couchdb-couchdb-1   standard                 19h

/project # kubectl --namespace kube-system get pv
NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS    CLAIM                                     STORAGECLASS   REASON    AGE
pvc-21aeec86-896a-11e8-a157-42010a800006   10Gi       RWO            Delete           Bound     gpii/database-storage-couchdb-couchdb-0   standard                 19h
pvc-21b3c7f5-896a-11e8-a157-42010a800006   10Gi       RWO            Delete           Bound     gpii/database-storage-couchdb-couchdb-1   standard                 19h
```
Only resource consumers (i.e. pods, pvcs, etc.) are bound to namespaces.

[the remark this answers is not in the document]
Comment (mrtyler) [reply gcp/modules/couchdb/main.tf]: Haha, of course. Good catch Sergey.

File: gcp/modules/couchdb/main.tf (hunk 9 above)
[the remark this answers is not in the document]
Comment (natarajaya) [reply gcp/modules/couchdb/main.tf]: We want to get all PVs claimed by CouchDB. And all CouchDB PVCs got unique names with `database-storage-couchdb` as common part. To apply this condition in [JSONPath filter](https://kubernetes.io/docs/reference/kubectl/jsonpath/) we need either regexp match (`=~`) or string function (substring, i.e. startsWith) and none of these available in kubectl jsonpath implementation.

[the remark this answers is not in the document]
Comment (natarajaya) [reply gcp/modules/couchdb/main.tf]: Done. What about stg/prd Terragrunt modules for CouchDB? I was about to add them as well.

File: gcp/live/dev/k8s/gpii/couchdb/terraform.tfvars (hunk 11 above)
Comment (natarajaya) [reply gcp/live/dev/k8s/gpii/couchdb/terraform.tfvars]: Ok, updated stg/prd couchdb Terragrunt modules.

**Comment by mrtyler** [comment]
Ah, yes! I forgot about stg and prd for a moment :).

Now this really LGTM. Ship it! I'm excited to have CouchDB with backups in GCP!
